From 39fbec005df2741978101e704167d73633f64c8a Mon Sep 17 00:00:00 2001
From: Chris Blake <chrisrblake93@gmail.com>
Date: Tue, 23 Jul 2019 12:45:30 -0500
Subject: [PATCH 01/21] arm64: Add dtb device info detection

This allows for the device name and serial to be loaded from the dtb
into /proc/cpuinfo
---
 arch/arm64/kernel/cpuinfo.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/arch/arm64/kernel/cpuinfo.c b/arch/arm64/kernel/cpuinfo.c
index 86136075ae41..1868da3590c0 100644
--- a/arch/arm64/kernel/cpuinfo.c
+++ b/arch/arm64/kernel/cpuinfo.c
@@ -17,6 +17,8 @@
 #include <linux/elf.h>
 #include <linux/init.h>
 #include <linux/kernel.h>
+#include <linux/of.h>
+#include <linux/of_fdt.h>
 #include <linux/personality.h>
 #include <linux/preempt.h>
 #include <linux/printk.h>
@@ -136,6 +138,10 @@ static int c_show(struct seq_file *m, void *v)
 {
 	int i, j;
 	bool compat = personality(current->personality) == PER_LINUX32;
+	const char *board_name;
+	const char *system_serial;
+	struct device_node *root;
+	int ret;
 
 	for_each_online_cpu(i) {
 		struct cpuinfo_arm64 *cpuinfo = &per_cpu(cpu_data, i);
@@ -187,6 +193,23 @@ static int c_show(struct seq_file *m, void *v)
 		seq_printf(m, "CPU revision\t: %d\n\n", MIDR_REVISION(midr));
 	}
 
+	/* load our system info from our DT */
+	root = of_find_node_by_path("/");
+	if (root) {
+		ret = of_property_read_string(root, "serial-number",
+					      &system_serial);
+		if (ret)
+			system_serial = "0000000000000000";
+
+		ret = of_property_read_string(root, "model",
+						      &board_name);
+		if (ret)
+			board_name = "Generic arm64 board";
+	}
+	seq_printf(m, "Hardware\t: %s\n", board_name);
+	seq_printf(m, "Revision\t: %s\n", "0000");
+	seq_printf(m, "Serial\t\t: %s\n", system_serial);
+
 	return 0;
 }
 
-- 
2.17.1

